gelf/reader_test.go

```diff
@@ -0,0 +1,219 @@
+	chunked1 := createMessage("b", ChunkSize+1)
+	chunkedBytes1 := prepareUdpMessages(t, chunked1)
+	//message that is sent in 3 chunks
+	chunked2 := createMessage("c", ChunkSize*2)
+	chunkedBytes2 := prepareUdpMessages(t, chunked2)
+	//message that is sent in 129 chunks
+	messageBytes3 := prepareUdpMessages(t, createMessage("d", chunkedDataLen*128))
+
+	tests := map[string]struct {
+		udpMessagesToSend [][]byte
+		expectedMessages  []Message
+		expectedErrors    []error
+	}{
+		"expected messages to be received even if chunked messages are received in parallel": {
+			udpMessagesToSend: [][]byte{chunkedBytes1[0], chunkedBytes2[0], chunkedBytes2[1], chunkedBytes1[1], chunkedBytes2[2]},
+			expectedMessages:  []Message{chunked1, chunked2},
+		},
+		"expected message to be received even if chunks are sent in any order": {
+			udpMessagesToSend: [][]byte{chunkedBytes2[1], chunkedBytes2[0], chunkedBytes2[2]},
+			expectedMessages:  []Message{chunked2},
+		},
+		"expected messages to be received even if not chunked message is sent between chunks": {
+			udpMessagesToSend: [][]byte{chunkedBytes1[0], notChunkedBytes[0], chunkedBytes1[1]},
+			expectedMessages:  []Message{chunked1, notChunked},
+		},
+		"expected not chunked messages to be received": {
+			udpMessagesToSend: [][]byte{notChunkedBytes[0], notChunkedBytes[0]},
+			expectedMessages:  []Message{notChunked, notChunked},
+		},
+		"expected error if chunked message length is not greater than 12 bytes": {
+			udpMessagesToSend: [][]byte{magicChunked},
+			expectedErrors:    []error{fmt.Errorf("chunked message size must be greather than 12")},
+		},
+		"expected error if message is split to more than 128 chunks": {
+			udpMessagesToSend: messageBytes3,
+			expectedErrors:    []error{fmt.Errorf("message must not be split into more than %v chunks", maxChunksCount)},
+		},
+	}
+
+	for name, data := range tests {
+		t.Run(name, func(t *testing.T) {
+			reader, err := NewReader(":0")
+			require.NoError(t, err)
+			done := make(chan bool)
+			defer reader.Close()
+			defer close(done)
+
+			addr, err := net.ResolveUDPAddr("udp", reader.Addr())
+			require.NoError(t, err)
+
+			conn, err := net.DialUDP("udp", nil, addr)
+			defer conn.Close()
+
+			require.NoError(t, err)
+
+			receivedMessages := make([]Message, 0, len(data.expectedMessages))
+			errors := make([]error, 0)
+			g, _ := errgroup.WithContext(context.Background())
+			messagesChannel := make(chan Message)
+			errorsChannel := make(chan error)
+			go func() {
+				defer func() {
+					if r := recover(); r != nil {
+						errorsChannel <- fmt.Errorf("panic during execution: %v", r)
+					}
+				}()
+				for {
+					select {
+					case <-done:
+						return
+					default:
+						msg, err := reader.ReadMessage()
+						if err != nil {
+							errorsChannel <- err
+						} else {
+							messagesChannel <- *msg
+						}
+					}
+				}
+			}()
+			g.Go(func() error {
+				for {
+					select {
+					case <-time.After(1 * time.Second):
```

Comment:
This might be flaky if the system is super busy and otherwise lengthen the `go test` time. Maybe you can create a channel that closes when all expected messages have arrived.

```suggestion
					case <-time.After(1 * time.Second):
```

Reply:
@simonswine yes, I see your point. but we do not have CI configured for this repository. Can we leave it as is? 
otherwise, it will be required to change the test logic completely

Reply:
I agree I don't think we need to add CI.

I still think the code is a bit too complex and the `time.After` unnecessary. I think you actually don't need more than a single go routine.

 I have basically butchered your code and I think that should do the same thing: https://github.com/grafana/go-gelf/compare/changed_reader_behavior...simonswine:20220421_simplify?expand=1

Reply:
done